File: 백준/Silver/2579. 계단 오르기/계단 오르기.cc
#include <iostream>

#define MAX 301
using namespace std;


int main()
{
    int N; // 계단 의 수
    cin >> N;
    
    int stair[MAX] = {0, };
    for(int i = 1; i <= N; i++){
        cin >> stair[i];
    }
    
    int dp[MAX] = {0, }; //  DP ARRAY
    
    dp[0] = 0;
    dp[1] = stair[1];
    dp[2] = stair[1] + stair[2];
    dp[3] = max(stair[1] + stair[3] , stair[2] + stair[3]);
    for(int i = 4; i <=N; i++){
        dp[i] = max(dp[i-2] + stair[i], dp[i-3] + stair[i-1] + stair[i]);
    }
    
    cout << dp[N];

    return 0;
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
n = int(input())
stair = [0]
for _ in range(n):
    stair.append(int(input()))

dp = [0] * (n + 1)
dp[1] = stair[1]
if n == 1:
    print(dp[1])
    exit()

dp[2] = stair[1] + stair[2]
if n == 2:
    print(dp[2])
    exit()


for i in range(3, n):
    dp[i] = max(stair[i] + stair[i - 1] + dp[i - 3], 
                stair[i] + dp[i-2], 
                stair[i -1])
print(max(stair[n] + dp[n - 2], stair[n] + stair[n - 1] + dp[n - 3]))